### ISSUE TITLE
training speed and the GPU usage

### ISSUE THREAD

**Issue by HanHenryOnlyFaceany** [ipis-mjkim/caueeg-ceednet#1]
Hello, I am running your model, the GPU I am using is 3060, and I am running resnet-18, but the training speed is very slow, the GPU usage is not high, and even the training day is less than a quarter of the total, And I want to know how the total samples are determined. It is 100,000 in tiny and 100 million in base-train. I hope to get your reply，Thanks
![image](https://github.com/ipis-mjkim/caueeg-ceednet/assets/75490877/d3ac1319-d663-4fc5-a8e6-5168ec89faa5)

**Comment by ipis-mjkim** [comment]
Hello @HanHenryOnlyFaceany ,

Thank you for raising your concerns. 
Let's address each one.

**First issue - Training speed and GPU usage**

There are a couple of potential remedies you might consider:

1. File format: In many cases, reading original `EDF` files can be slow and cause a training bottleneck. If so, you can convert the file format using the following command.
```bash
python ./datasets/convert_file_format.py  # This command only needs to be run once and takes a few minutes
python run_train.py data=caueeg-dementia model=1D-ResNet-18 train=base_train ++data.file_format=feather
```
2. NVIDIA driver and CUDA installation: Please ensure that your NVIDIA driver, CUDA, and PyTorch are properly installed. To confirm this, you can run the example code provided in this [link](https://github.com/pytorch/examples/tree/main/vision_transformer). While executing this code, monitor the GPU usage using `nvidia-smi`; it should be high (in my experience, it exceeds 80%).

If these steps don't resolve the issue, please provide the following information to help me assist you better:

1. The Python command you executed.
2. The output of the `pip list` command.
3. The output of the `nvidia-smi` command during training.

**Second issue - Number of training samples**
I've set the default option to 1M and used a smaller version (100K) for debugging. However, it's important to mention that the number 1M was chosen as a sufficiently large number for our use case, and smaller numbers should also work appropriately in most scenarios. You can adjust this according to your requirements.

I hope this is helpful.
Please let me know if you have further questions.

**Comment by HanHenryOnlyFaceany** [comment]
**Thank you for your prompt reply.**
**I still have some doubts and would like to ask you for advice.**

**First issue - Training speed and GPU usage**

Based on your suggestion, I did some work,Here are a few situations I've encountered:

  1.File format: I used the file format of feather and adjusted the training to tiny-train, the following is the python command I executed.

![image-20230707112026751](https://github.com/ipis-mjkim/caueeg-ceednet/assets/75490877/8cc70896-8d7d-4ec3-8906-f1d432775d8e)

At the same time I monitored my GPU usage as follows:

![image-20230707111507192](E:\每日笔记\有关caueeg的一些thing.assets\image-20230707111507192.png)

I observed that the usage rate of cuda is oscillating. I think it is the speed mismatch between cpu and GPU, but the highest point does not exceed 40%. I want to know if this is normal.

  2.NVIDIA driver and CUDA installation: I ran the sample code based on your suggestion, and the NVIDIA drivers, CUDA and PyTorch are installed correctly, as shown in the image below:

![image-20230707102220898](E:\每日笔记\有关caueeg的一些thing.assets\image-20230707102220898.png)



**Second issue - Number of Total samples**
I don't understand why the number of training samples can be modified by itself. Isn't the number set according to the size of the data set?I try to increase the number of samples of tiny-train or reduce the number of samples of base-train. I am experimenting with the number of samples, what will affect the accuracy?I found that when I modify the total samples, the total number of samples in my confusion matrix does not change, but the accuracy does improve. I want to know what the total samples represent，Is the size of a sample (20, 2000)? We divide the EEG of a patient into multiple samples for training?

**Third issue - Questions about training batches**

I don't know how many rounds we have trained the training data in the end. Is there only one round of training?



**Regarding the above two questions (data=caueeg-dementia):**

 I think the number of complete training samples in one round is 100,000, and if I set the total samples to 1,000,000, it means that I want to perform ten rounds of training on the entire training set, epochs=10 , is this understanding correct?

Thank you for your answer

**Comment by HanHenryOnlyFaceany** [comment]
**Thank you for your prompt reply.**
**I still have some doubts and would like to ask you for advice.**

**First issue - Training speed and GPU usage**

Based on your suggestion, I did some work,Here are a few situations I've encountered:

  1.File format: I used the file format of feather and adjusted the training to tiny-train, the following is the python command I executed.

![image-20230707112026751](https://github.com/ipis-mjkim/caueeg-ceednet/assets/75490877/a521b63f-dc83-4032-aa11-cff52fe10b62)

At the same time I monitored my GPU usage as follows:
![image-20230707111507192](https://github.com/ipis-mjkim/caueeg-ceednet/assets/75490877/fc453ab4-14bb-45e5-bd18-4f8ca8917310)

I observed that the usage rate of cuda is oscillating. I think it is the speed mismatch between cpu and GPU, but the highest point does not exceed 40%. I want to know if this is normal.

  2.NVIDIA driver and CUDA installation: I ran the sample code based on your suggestion, and the NVIDIA drivers, CUDA and PyTorch are installed correctly, as shown in the image below:

![image-20230707102220898](https://github.com/ipis-mjkim/caueeg-ceednet/assets/75490877/2d4e5b49-9943-4b96-af4f-b10bbe45307f)



**Second issue - Number of Total samples**
I don't understand why the number of training samples can be modified by itself. Isn't the number set according to the size of the data set?I try to increase the number of samples of tiny-train or reduce the number of samples of base-train. I am experimenting with the number of samples, what will affect the accuracy?I found that when I modify the total samples, the total number of samples in my confusion matrix does not change, but the accuracy does improve. I want to know what the total samples represent，Is the size of a sample (20, 2000)? We divide the EEG of a patient into multiple samples for training?

**Third issue - Questions about training batches**

I don't know how many rounds we have trained the training data in the end. Is there only one round of training?



**Regarding the above two questions (data=caueeg-dementia):**

 I think the number of complete training samples in one round is 100,000, and if I set the total samples to 1,000,000, it means that I want to perform ten rounds of training on the entire training set, epochs=10 , is this understanding correct?

Thank you for your answer

**Comment by HanHenryOnlyFaceany** [comment]
num_workers = 0  # A number other than 0 causes an error 
Why can't num_workers be set to 4 or 8? This improves GPU utilization

**Comment by ipis-mjkim** [comment]
Thank you for your updates.
I will address each one.


**First issue - Training speed and GPU usage**

I understand you converted the file format and installed the NVIDIA and CUDA setups.
Based on your screenshots, I suggest some trials to boost the training speed.

The storage capacity of each EEG recording is relatively large (amounts to 5-20 MB).
The main bottleneck during training may be the data transfer between the SSD/HDD and the CPU/GPU.
To reduce the bottleneck between storage and CPU/GPU, we can consider several options, including (1) Preloading the whole dataset onto RAM, (2) Using other file formats faster than Feather, (3) Pre-cropping each EEG recording into multiple small segments, and (4) Using `crop_multiple` option.
Options (1), (2), and (3) require modifying the `CauEegDataset._read_signal()` function. 
(3) has an additional demerit: we cannot use intervals between small segments for training.
For option (4), we can add the command such as `++data.crop_multiple=2`.
With this option (4), we crop multiple segments from a single EEG recording and utilize them for training at every iteration.
This helps us set the `minibatch` value larger than 950 (training data size).
Notably, the specified `minibatch` value should be multiple of the `crop_multiple` option.

As you described, `num_workers` can also be tuned depending on the environment.


**Second issue - Number of training samples**
**Third issue - Questions about training batches**
**Fourth issue - Regarding the above two questions (data=caueeg-dementia)**
In this repository, I used the term `total_samples` as the substitute for the "total epochs."
The exact meaning of `total_samples` is the number of training data the model met during training, not the training data size.
In other words, `total_samples` = (# epochs) x  (# training data = 950).
We apologize for the confusion.


I hope these are helpful.
Please let me know if you have further questions or if there are updates.
